### ISSUE TITLE
False positive RedundantCondition for an array field, even if array is reassigned

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#397]
Observed: RedundantCondition on $data['field']
Expected: No issues emitted. The whole array changed, so information about that array's fields should be invalidated.
This is probably infrequent.

Also, if I remove the rand() &&, the issue is not emitted.

```php
<?php

function get_new_array() : array {
    return ['field' => true];
}

function array_check() : void {
    $data = ['field' => false];
    while (rand(0, 1) > 0 && !$data['field']) {  // See RedundantCondition for $data['field']
        $data = get_new_array(); // Changing this to a literal would get rid of RedundantCondition
    }
}
```

**Comment by muglug** [comment]
Thanks – fixed in https://github.com/vimeo/psalm/commit/8ca5fbefea253e6e54dab6556fdefbc203369c29 because nested logical clauses (i.e. on array keys and properties) weren't being discarded when assigning to their roots
